stats/opentelemetry/internal/grpc_trace_bin_propagator.go

```diff
@@ -0,0 +1,132 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package internal
+
+import (
+	"context"
+	"encoding/base64"
+
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+	otelinternaltracing "google.golang.org/grpc/stats/opentelemetry/internal/tracing"
+)
+
+// TODO: Move out of internal as part of open telemetry API
+
+// GRPCTraceBinPropagator is TextMapPropagator to propagate cross-cutting
+// concerns as both text and binary key-value pairs within a carrier that
+// travels in-band across process boundaries.
+type GRPCTraceBinPropagator struct{}
+
+// Inject set cross-cutting concerns from the Context into the carrier.
+//
+// If carrier is carrier.CustomMapCarrier then SetBinary (fast path) is used,
+// otherwise Set (slow path) with encoding is used.
+func (p GRPCTraceBinPropagator) Inject(ctx context.Context, carrier propagation.TextMapCarrier) {
+	span := trace.SpanFromContext(ctx)
+	if !span.SpanContext().IsValid() {
+		return
+	}
+
+	binaryData := Binary(span.SpanContext())
+	if binaryData == nil {
+		return
+	}
+
+	if customCarrier, ok := carrier.(otelinternaltracing.CustomCarrier); ok {
+		customCarrier.SetBinary(binaryData) // fast path: set the binary data without encoding
+	} else {
+		carrier.Set(otelinternaltracing.GRPCTraceBinHeaderKey, base64.StdEncoding.EncodeToString(binaryData)) // slow path: set the binary data with encoding
+	}
+}
+
+// Extract reads cross-cutting concerns from the carrier into a Context.
+//
+// If carrier is carrier.CustomCarrier then GetBinary (fast path) is used,
+// otherwise Get (slow path) with decoding is used.
+func (p GRPCTraceBinPropagator) Extract(ctx context.Context, carrier propagation.TextMapCarrier) context.Context {
+	var binaryData []byte
+
+	if customCarrier, ok := carrier.(otelinternaltracing.CustomCarrier); ok {
+		binaryData, _ = customCarrier.GetBinary()
+	} else {
+		binaryData, _ = base64.StdEncoding.DecodeString(carrier.Get(otelinternaltracing.GRPCTraceBinHeaderKey))
+	}
+	if binaryData == nil {
+		return ctx
+	}
+
+	spanContext, ok := FromBinary([]byte(binaryData))
+	if !ok {
+		return ctx
+	}
+
+	return trace.ContextWithRemoteSpanContext(ctx, spanContext)
+}
+
+// Fields returns the keys whose values are set with Inject.
+//
+// GRPCTraceBinPropagator will only have `grpc-trace-bin` field.
```

Comment:
What does this mean?

Reply:
Added more details in comment. Let me know its clear. Basically the `propagation.TextMapCarrier` interface which `GRPCTraceBinPropagator` implements has `Keys()` which returns the keys that are set in propagator. However, `GRPCTraceBinPropagator` only has one header key `grpc-trace-bin`